### PR TITLE
chore(Python): add Rabin Karp algorithm

### PR DESCRIPTION
**Pull request by kartiknad**
**PR Checklist:**
- [x] My submission is formatted according to the guidelines in the [contributing guide](https://github.com/MakeContributions/DSA#contribution-guidelines)
- [x] My addition is on refer on the language README.md file
- [x] My addition does not have a spelling problem
- [x] My submission has a proper and user-friendly description of the algorithm
- [x] My submission has the time complexity of the algorithm
- [x] My submission has sample input-output of the program (NOT FOR PYTHON)

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] New algorithm
- [ ] Optimization in previous algorithms
- [ ] Code style update
- [ ] Refactor
- [ ] Documentation
- [ ] Other, please describe:

**Briefly describe the changes in this PR**

Added a new algorithm, Rabin-Karp pattern matching algorithm implemented in python

**Other information:**

### CODE REVIEW

File: algorithms/Python/strings/rabin-karp-algorithm.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    
+    if not found: # If pattern not found in text
+        pos.append(-1)
+        
+    #Printing results
+    if pos[0]==-1:
+        print("Pattern not found")
+    else:
+        for i in pos:
+            print("Pattern found at pos: {}".format(i))
```
Comment (Arsenic-ATG) [inline algorithms/Python/strings/rabin-karp-algorithm.py:90]: The pyupgrade test is most probably failing due to use use of old style string formatting here, and recommends to use f-strings instead 

```suggestion
            print(f"Pattern found at pos: {i}")
```

File: algorithms/Python/strings/rabin-karp-algorithm.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
+'''
+String pattern matching algorithm which performs efficiently for large text and patterns
+
+Algorithm: 
+Rabin Karp works on the concept of hashing. The if the substring of the given text 
+is same as the pattern then the corresponding hash value should also be same. Exploting this idea 
+and designing a hash function which can be computed in O(m) time for both pattern and initial window 
+of text. The subsequent  window each will require only O(1) time. And we slide the window n-m times 
+after the initial window. Therefore the overall complexity of calculating hash funtion for text is O(n-m+1)
+Once the hash value matches, the underlying string is again checked with pattern for matching
```
Comment (Arsenic-ATG) [inline algorithms/Python/strings/rabin-karp-algorithm.py:10]:
```suggestion
Algorithm: 
Rabin Karp works on the concept of hashing. If the substring of the given text 
is same as the pattern then the corresponding hash value should also be same. Exploiting this idea 
and designing a hash function which can be computed in O(m) time for both pattern and initial window 
of text. The subsequent  window each will require only O(1) time. And we slide the window n-m times 
after the initial window. Therefore the overall complexity of calculating hash funtion for text is O(n-m+1)
Once the hash value matches, the underlying string is again checked with pattern for matching
```

File: algorithms/Python/strings/rabin-karp-algorithm.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    
+            T: string
+               The string where the pattern needs to be searched
+
+            P: string 
+               The pattern to be searched
+
+            q: int
+               An appropriately chosen prime number based on length of input strings
+               The higher the prime number, the lower the collisiona nd spurious hits
```
Comment (Arsenic-ATG) [inline algorithms/Python/strings/rabin-karp-algorithm.py:32]:
```suggestion
               The higher the prime number, the lower the collision and spurious hits
```

**Review by Arsenic-ATG — CHANGES_REQUESTED** [review]
The suggested changes should make the PR pass all the pre-tests.

**Comment by ming-tsai** [comment]
Hi @kartiknad,
Any update?

**Comment by kartiknad** [comment]
> Hi @kartiknad, Any update?

The suggested changes have been made

File: algorithms/Python/strings/rabin-karp-algorithm.py (hunk 1 above)
Comment (kartiknad) [reply algorithms/Python/strings/rabin-karp-algorithm.py]: Thanks for the suggestions, I have made the changes

**Review by Arsenic-ATG — APPROVED** [review]
My last commit involve fixing some more nit-picks detected by pyupgrade and codespell checks. 

the PR should be good to go now.

**Pull request merged**